test/extended/operators/workload_guarantees.go

```diff
@@ -0,0 +1,269 @@
+		knownBugs := map[string]string{
+			"HA/Deployment/openshift-image-registry/image-registry": "https://bugzilla.redhat.com/show_bug.cgi?id=1945387",
+			"HA/Deployment/openshift-ingress/router-default":        "https://bugzilla.redhat.com/show_bug.cgi?id=1945393",
+			"HA/Deployment/openshift-monitoring/prometheus-adapter": "https://bugzilla.redhat.com/show_bug.cgi?id=1948711",
+			"HA/Deployment/openshift-monitoring/thanos-querier":     "https://bugzilla.redhat.com/show_bug.cgi?id=1948711",
+			"HA/StatefulSet/openshift-monitoring/prometheus-k8s":    "https://bugzilla.redhat.com/show_bug.cgi?id=1949262",
+			"HA/StatefulSet/openshift-monitoring/alertmanager-main": "https://bugzilla.redhat.com/show_bug.cgi?id=1949262",
+		}
+
+		masterReq, err := labels.NewRequirement("node-role.kubernetes.io/master", selection.Exists, []string{})
+		if err != nil {
+			e2e.Failf("unable to build label requirement: %v", err)
+		}
+
+		var knownBroken []string
+		var invalidWorkloads []string
+		var replicas int32
+		for _, workload := range items {
+			var meta *metav1.ObjectMeta
+			var template *corev1.PodTemplateSpec
+			switch t := workload.(type) {
+			case *appsv1.Deployment:
+				t.Kind = "Deployment"
+				if t.Spec.Replicas == nil {
+					continue
+				}
+				meta = &t.ObjectMeta
+				template = &t.Spec.Template
+				replicas = *t.Spec.Replicas
+			case *appsv1.StatefulSet:
+				t.Kind = "StatefulSet"
+				if t.Spec.Replicas == nil {
+					continue
+				}
+				meta = &t.ObjectMeta
+				template = &t.Spec.Template
+				replicas = *t.Spec.Replicas
+			default:
+				panic("not an object")
+			}
+
+			if !strings.HasPrefix(meta.Namespace, "openshift-") {
+				continue
+			}
+
+			// exclude workloads that are master specific
+			if template.Spec.NodeSelector != nil {
+				if masterReq.Matches(labels.Set(template.Spec.NodeSelector)) {
+					continue
+				}
+			}
+			var toleratesMaster bool
+			for _, tolerates := range template.Spec.Tolerations {
+				if tolerates.Key == "node-role.kubernetes.io/master" && tolerates.Operator == corev1.TolerationOpExists {
+					toleratesMaster = true
+				}
+			}
+			if toleratesMaster {
+				continue
+			}
+
+			key := fmt.Sprintf("%s/%s/%s", workload.GetObjectKind().GroupVersionKind().Kind, meta.Namespace, meta.Name)
+
+			if inHAMode {
+				if replicas == 1 {
+					// TODO: We need to check if this is a misconfiguration.
+					e2e.Logf("The %s component has a single replica when OpenShift is configured to be HA", key)
+					continue
+				}
+				workloadStatus, isInvalidWorkload := isInvalidWorkloadInHAMode(knownBugs, key, workload, template)
+				if isInvalidWorkload && workloadStatus != "" {
+					invalidWorkloads = append(invalidWorkloads, workloadStatus)
+				} else if !isInvalidWorkload && workloadStatus != "" {
+					knownBroken = append(knownBroken, workloadStatus)
+				}
+			} else if !inHAMode {
+				// For SNO cases
+				nonHAPrefix := "non-HA/"
```

Comment:
@romfreiman  @dhellmann - Can you please review this block to see if it covers all the SNO cases?

Reply:
I don't see any components listed in `knownBugs` with a key like the one built here. We know that some components still do not reduce their replica count for single-node deployments. The package server from OLM, for example (https://issues.redhat.com/browse/OLM-2017).

Rather than adding the job and blocking work for a bunch of teams, could you look at a single node deployment to see if there are other components that should be added to the list?

Reply:
The main goal here is to identify the components that don't confirm to single replica and affinities and create bugs for them. I can access a SNO via CI, if you can assign someone to work with me in merging this test it'll be great. I especially want review from you folks on this block- https://github.com/openshift/origin/pull/26067/files#diff-3e1d6d8499bb3600ca701074807127cdecbd73803989c6178c026e3d2b48ad6cR159. @romfreiman  has mentioned that you are already covering single replica whitelist here - https://github.com/openshift/origin/blob/master/test/extended/single_node/topology.go#L102